server/bin/pbench-cull-unpacked-tarballs.py

```diff
@@ -0,0 +1,457 @@
+        return 3
+
+    if not os.path.isdir(users_p):
+        logger.error(
+            "The configured USERS directory, {}, is not a valid directory", users
+        )
+        return 4
+
+    # Fetch the configured maximum number of days a tar can remain "unpacked"
+    # in the INCOMING tree.
+    try:
+        max_unpacked_age = config.conf.get("pbench-server", "max-unpacked-age")
+    except NoOptionError as e:
+        logger.error(f"{e}")
+        return 5
+    try:
+        max_unpacked_age = int(max_unpacked_age)
+    except Exception:
+        logger.error("Bad maximum unpacked age, {}", max_unpacked_age)
+        return 6
+
+    # First phase is to find all the tar balls which are beyond the max
+    # unpacked age, and which still have an unpacked directory in INCOMING.
+    if config._ref_datetime is not None:
+        try:
+            curr_dt = config._ref_datetime
+        except Exception:
+            # Ignore bad dates from test environment.
+            curr_dt = datetime.utcnow()
+    else:
+        curr_dt = datetime.utcnow()
+
+    _msg = "Culling unpacked tar balls {} days older than {}"
+    if options.dry_run:
+        print(_msg.format(max_unpacked_age, curr_dt.strftime(_STD_DATETIME_FMT)))
+    else:
+        logger.debug(_msg, max_unpacked_age, curr_dt.strftime(_STD_DATETIME_FMT))
+
+    actions_taken = []
+    errors = 0
+    start = pbench._time()
+    for tb_incoming_dir, controller_name in gen_list_unpacked_aged(
+        incoming_p, archive_p, curr_dt, max_unpacked_age
+    ):
+        errors, _actions_taken, duration = remove_unpacked(
+            tb_incoming_dir,
+            controller_name,
+            results_p,
+            users_p,
+            logger,
+            options.dry_run,
+        )
+        unpacked_dir_name = os.path.basename(tb_incoming_dir)
+        act_name = os.path.join(controller_name, unpacked_dir_name)
+        actions_taken.append((_actions_taken, act_name, errors, duration))
+        if errors > 0:
+            # Stop any further unpacked tar ball removal if an error is
+            # encountered.
+            break
+    end = pbench._time()
+
+    if options.dry_run:
+        # We don't emit a report for a dry run.
+        return errors
+
+    # Generate the ${TOP}/public_html prefix so we can strip it from the
+    # various targets in the report.
+    public_html = os.path.realpath(os.path.join(config.TOP, "public_html"))
+
+    # Write the actions taken into a report file.
+    with tempfile.NamedTemporaryFile(
+        mode="w+t", prefix=f"{_NAME_}.", suffix=".report", dir=config.TMP
+    ) as tfp:
+        duration = end - start
+        total = len(actions_taken)
+        print(
+            f"Culled {total:d} unpacked tar balls ({errors:d} errors)"
+            f" in {duration:0.2f} secs",
+            file=tfp,
+        )
+        if total > 0:
+            print(f"\nActions Taken:", file=tfp)
+        for (act_l, act_n, act_e, act_d) in actions_taken:
```

Comment:
It might be nice to document the structure of `actions_taken` somewhere: I had to jump around to figure it out and if somebody modifies this script in the future, I think they will appreciate the guidance!

Reply:
I'll fix this up using a class.